libdrgn/arch_s390.c

```diff
@@ -0,0 +1,327 @@
+	if (!regs2)
+		goto out;
+
+	uint8_t *ptregs = frame + frame_size;
+	pswa = (uint64_t *)(ptregs + (psw_offset + addr_offset) / 8);
+	pswm = (uint64_t *)(ptregs + (psw_offset + mask_offset) / 8);
+
+        uint64_t *gprs = (uint64_t *)(ptregs + gprs_offset / 8);
+	drgn_register_state_set_range_from_buffer(regs2, r0, r15, gprs);
+	drgn_register_state_set_from_buffer(regs2, pswa, pswa);
+	drgn_register_state_set_from_buffer(regs2, pswm, pswm);
+	drgn_register_state_set_pc_from_register(prog, regs2, pswa);
+	drgn_register_state_set_from_buffer(regs2, bc, &gprs[15]);
+
+	*ret = regs2;
+
+	if (*pswm & (1UL << 48)) {
+		/* don't follow to userspace */
+	        return &drgn_stop;
+	}
+	return NULL;
+out:
+	free(frame);
+	return &drgn_enomem;
+}
+
+static struct drgn_error *
+fallback_unwind_s390(struct drgn_program *prog,
+		     struct drgn_register_state *regs,
+		     struct drgn_register_state **ret)
+{
+	struct optional_uint64 bc = drgn_register_state_get_u64(prog, regs, bc);
+	struct drgn_qualified_type frame_type;
+	struct drgn_type_member *member;
+	struct drgn_error *err;
+	uint64_t bit_offset;
+	uint8_t frame[256];
+
+	if (!bc.value)
+		return unwind_check_pt_regs(prog, regs, ret);
+
+	err = drgn_program_read_memory(prog, frame, bc.value, sizeof(frame), false);
+	if (err)
+		return err;
+
+	err = drgn_program_find_type(prog, "struct stack_frame", NULL, &frame_type);
+	if (err)
+		return err;
+
+	err = drgn_type_find_member(frame_type.type, "gprs", &member, &bit_offset);
+	if (err)
+		return err;
+
+	drgn_register_state_set_range_from_buffer(*ret, r6, r15, frame + bit_offset / 8);
+	drgn_register_state_set_pc_from_register(prog, *ret, r14);
+
+	err = drgn_type_find_member(frame_type.type, "back_chain", &member, &bit_offset);
+	if (err)
+		return err;
+
+	drgn_register_state_set_from_u64(prog, *ret, bc, *(uint64_t *)(frame + bit_offset / 8));
+	return NULL;
+}
+
+/*
+ * The linux kernel saves the callee-saved registers in a normal
+ * stack frame and saves this address to struct task_struct.thread.ksp.
+ * See __switch_to() in arch/s390/kernel/entry.S
+ */
+static struct drgn_error *
+linux_kernel_get_initial_registers_s390(const struct drgn_object *task_obj,
+					   struct drgn_register_state **ret)
+{
+	struct drgn_program *prog = drgn_object_program(task_obj);
+	struct drgn_register_state *regs, *regs2;
+	struct drgn_qualified_type frame_type;
+	struct drgn_object ctx;
+	struct drgn_error *err;
+	uint64_t ksp;
+
+	drgn_object_init(&ctx, prog);
+
+	err = drgn_object_member_dereference(&ctx, task_obj, "thread");
+	if (err)
+		goto out;
+
+	err = drgn_object_member(&ctx, &ctx, "ksp");
+	if (err)
+		goto out;
+
+	err = drgn_object_read_unsigned(&ctx, &ksp);
+	if (err)
+		goto out;
+
+	regs = drgn_register_state_create(bc, false);
+	if (!regs)
+		return &drgn_enomem;
+
+	drgn_register_state_set_from_u64(prog, regs, bc, ksp);
+
+	regs2 = drgn_register_state_create(bc, false);
```

Comment:
You can probably avoid creating a fake frame by having this function and `fallback_unwind_s390()` both call another function that takes the backchain as a `uint64_t`. However, I suspect that given my point above about only recovering `r14` and `r15` for the fallback case, I suspect that these two versions will be different enough that you won't need to share anything anyways.